ROX-19004: NVD CVSS data updating pipeline ADR (#7765)
stackrox/stackrox: scanner/decisions/0004-scannerv4-read-vulneralbilities.md

# 0004 - Scanner V4 NVD CVSS enrichment data updating pipeline

- **Author(s):** Yi Li
- **Created:** [2023-09-11]

## Context

We are in the process of transitioning the StackRox Scanner to utilize ClairCore as its primary scanning engine. Our objective is twofold: firstly, to allow Central to fetch NVD CVSS data, and secondly, to ensure the Scanner V4 seamlessly connecting with Central, triggering data updates and enriching vulnerabilities with CVSS data. Throughout this pipeline, the Claircore enricher will play a pivotal role. It will manage tasks including downloading CVSS data from Google storage, parsing the information, and updating the data accordingly for Scanner V4

> ## Decision

Reviewer: I think the language of the ADR becomes easier to read and understand if you use imperative language rather than descriptive. One example for your reference: "The CVSS Updater will reside in Central" instead of "The CVSS Updater resides within Central".

> 

Author: 1. Scanner is starting, no CVSS has been pulled before -> download JSON always happens.
2. Scanner has donwloaded it before, but timestamp in storage is older (not modififer) -> download JSON don't happen.
3. Scanner has downlodaded it before, but timstamp in storage is newer (modified) -> dowload WILL happen.

Author: Add how we download the json.gz for each year. Are we downloading the whole repo or send 20 http requests.

Author: We are going to have the CVSSUpdater and it will pull CVSS data from google storage and will generate a consolidated enrichment bundle in central as a goRoutine.

> The CVSS Updater resides within Central, fetching CVSS data from Google storage upon request. This updater then stores the data in Central's file system. Throughout this download and save process, the Claircore CVSS enricher plays a significant role.

Reviewer: > This updater then stores the data in Central's file system

I remember the storage team moving these definitions to the DB in another PR some time ago. Can you confirm if the decision to use the FS won't be a regression?

Reviewer: > Throughout this download and save process, the Claircore CVSS enricher plays a significant role.

I apologize, but I'm having a bit of trouble understanding what you meant here. Could you please provide more information or clarification?

> The CVSS Updater resides within Central, fetching CVSS data from Google storage upon request. This updater then stores the data in Central's file system. Throughout this download and save process, the Claircore CVSS enricher plays a significant role.

Reviewer: > The CVSS Updater resides within Central, fetching CVSS data from Google storage upon request.

Perhaps it might be worth considering having it pull updates from the Storage on a regular basis, rather than upon request. This would allow us to respond with a "not modified" response in the CVSS Updater endpoint in Central, and eliminate the need to contact the storage every time. I can see how this approach could be particularly helpful, especially since updates to the storage content don't occur frequently.

CVSS Updater handler and singleton in central provides http URL for sensor and scanner to connect with and send the data retrieving request. Most importantly it serves with NVD data upon those requests.

> In Scanner V4, the component known as the 'data retriever' (previously referred to as the 'updater' in Scanner V2). However, its name 'updater' might be confused with the 'updater' that operates as a GitHub Action workflow in the Scanner V4 context. This data retriever in Scanner V4 communicates with the CVSS data retrieval endpoint, initiating data downloads in central and ensuring access to the latest CVSS data.

Reviewer: In the Secured Cluster, the Scanner cannot reach Central directly. The sensor needs to proxy those calls in some way.

Author: This CVSS data retriever in Scanner V4 matcher communicates with the CVSS data retrieval endpoint in Central.

### Central CVSS data retrieving Endpoints

> 1. `/api/extensions/scannerdefinitions`: is configured to handle GET requests, which are processed by the CVSS Updater handler in central.

Reviewer: Seems like you are deciding to use the same endpoint to retrieve the NVD CVSS data. Can you clarify a bit how the endpoint will support both previous data (Scanner V2's definitions, etc.) and the new one?

Author: /api/extensions/scannerdefinitions => /api/extensions/scanner-v4/definitions

## Consequences

> This document focuses exclusively on the pipeline for updating CVSS enrichment data. However, the design and structure of this pipeline could also be relevant for the 'repo to cpe' data and 'vulnerabilities' data pipelines. Thus, the architecture and principles underlying the CVSS data update process may potentially be generalized and adopted for all three scenarios in future iterations

Reviewer: We need to clarify how it can be reused for updating the other data specified (the mappings and vulnerability definitions). It is unclear if they are handled the same as NVD CVSS in Central and Scanner to generalize that the whole pipeline can be reused for them or only parts of the pipeline.

Author: We are not making any decisions for the updating pipeline of other data at this moment.

Author: GoRoutine will isolate this pipeline from Central, so it will not panic Central.

Author: Our own enricher consumes 150MB of memory 
Claircore enricher consumes 4.5GB of memory

# 0004 - Scanner V4 NVD CVSS enrichment data updating pipeline

> - **Author(s):** Yi Li

Reviewer: ```suggestion
- **Author(s):** Yi Li <yli3@redhat.com>, J. Victor Martins <jvdm@sdf.org>
```

- **Created:** [2023-09-11]

## Context

> We are currently transitioning the StackRox Scanner to use ClairCore as its primary scanning engine. Our goals are twofold:

Reviewer: ```suggestion
We are transitioning the StackRox Scanner to ClairCore as its primary scanning engine. To enrich vulnerabilities with CVSS information, ClairCore pulls CVSS scores from NVD during the Vulnerability Matching. However, Scanner V4 should not contact external endpoints other than registries and only contact Central for external data.

Our goal then are twofold:
```

1. Enable Central to retrieve NVD CVSS data and consolidate all CVSS data that contains available CVSS v3 scores.
2. Ensure seamless integration between Scanner V4 and Central, for enriching vulnerabilities with CVSS data.

> 
> Throughout this workflow, a new CVSS Updater equipped by a novel enricher in Central will download and consolidate the CVSS data, then creating a JSON bundle. 
> 
> ClairCore will be instrumental in scanner V4 data retriever (will be explained in detail below) while requesting this CVSS data bundle from central, handling tasks such as downloading, parsing, and updating CVSS data for Scanner V4.
> 
> ## Decision

Reviewer: ```suggestion

## Decision

Central will have a new NVD CVSS Updater equipped by a novel enricher, based on ClairCore. It will download and consolidate the NVD CVSS data in a JSON bundle. 

ClairCore will also be used in Scanner V4 CVSS data retriever (will be explained in detail below). It will require the NVD CVSS data bundle stored in Central. Central will handle the task of downloading, parsing, and updating the NVD CVSS data bundle for Scanner V4 using ClairCore.
```

> The CVSS Updater will be housed within Central, where it will retrieve CVSS data from Google storage when needed. It is designed to make a single HTTP call for each yearly NVD data bundle, with the earliest data tracing back to 2002. The updater will also compare the sha256 from the CVSS meta file for each year to determine if any data can be bypassed.

Reviewer: ```suggestion
The NVD CVSS Updater will retrieve CVSS data from Google storage at a configurable interval. It will make a single HTTP call for each yearly NVD data bundle, with the earliest data tracing back to 2002, and the latest to the current year. The updater will also compare the sha256 from the CVSS meta file for each year to detect and ignore corrupted data.
```

> This Updater will consolidate the data that only contains valid CVSS v3 scores, generate a json file and store it as a zip file in Central's file system. This compressed file is typically around 50 MB in size.

Reviewer 2: Per our talk today, I will suggest not to say Central's "file system", instead saying in Central's storage etc so that you can have the room on decision where you really want to store the file.

Reviewer: @daynewlee I don't see an immediate need for storing the NVD security data in persistent storage. If Central restarts (deployed, crashed, etc.), having it wiped out is OK because we would pull the data from Google Storage as soon as Central is ready. There is also the fact that it consumes less memory. And I think we can move to persistent later on if needed.

Nonetheless, Central still has Offline Mode. We should leave the door open to support that. We should not do anything right now that could prevent the customer from updating the feed manually. I don't think using `blob.Snapshot` would be a problem, but worth mentioning.

> Using a new CVSS enricher, the updater operates as a GoRoutine, set to refresh the CVSS data bundle every 4 hours. By leveraging GoRoutine, this pipeline runs independently from Central, ensuring that any failures won't disrupt Central.

Reviewer: ```suggestion
To generate the zip file, the CVSS updater will use a custom ClairCore CVSS Enricher.

The CVSS updater will operate as a GoRoutine, set to refresh the NVD CVSS data bundle at a configurable interval (default is 4 hours). By leveraging GoRoutine, this pipeline runs independently from Central, ensuring that any failures won't disrupt Central.
```

> The CVSS Updater handler, paired with a singleton in Central, offers an HTTP handler for both the sensor and scanner, facilitating data retrieval requests, fetching the existing data bundle and delivering consolidated NVD CVSS data via a http URL.

Reviewer: We don't need NVD CVSS data in Sensor, we don't have to mention it.

```suggestion
The CVSS Updater handler, paired with a singleton in Central, will offer an HTTP handler so Scanner can retrieve the consolidated NVD CVSS data.
```

> Within Scanner V4, there's a component named as 'data retriever', which in Scanner V2 was called the 'updater'. This name, 'updater', could be mistaken for the data updating GitHub action workflow in the Scanner V4 context. This CVSS data retriever in Scanner V4 communicates directly with Central's CVSS data retrieval endpoint.

Reviewer: ```suggestion
Within Scanner V4, there's a component named as 'data retriever', which in Scanner V2 was called the 'updater'. We will rename to retrieve because the name, 'updater', could be mistaken for the data updating GitHub action workflow in the Scanner V4 context. This CVSS data retriever in Scanner V4 communicates directly with Central's NVD CVSS data retrieval endpoint.
```

### Central CVSS data retrieving Endpoints

> 1. `/api/extensions/scanner-v4/definitions`: is configured to handle GET requests, which are processed by the CVSS Updater handler in central.

Reviewer: ```suggestion
1. `/api/extensions/scanner-v4/definitions`: is configured to handle GET requests, which are processed by the NVD CVSS Updater handler in central.
```

## Consequences

This document focuses exclusively on the pipeline for updating CVSS enrichment data. 

We are not making any decisions for the updating pipeline of other data (vulneralbility data and repo to cpe mapping data) at this moment.

A novel CVSS enricher mentioned above will be introduced and it will be equipped with an NVD CVSS JSON data parsing feature for the CVSS updater's use. 

> Unlike the Claircore CVSS enricher (which can also be utilized as a library/tool in Central), which loads JSON into memory before parsing, this new enricher parses JSON as a stream. The data processing and consolidation managed by this novel enricher utilize approximately 130MB of memory.

Reviewer: ```suggestion
Unlike the Claircore CVSS enricher (which can also be utilized as a library/tool in Central), which loads JSON into memory before parsing, this new enricher parses JSON as a stream. The data processing and consolidation managed by this novel enricher utilize approximately 130MB of memory.

We are including ClairCore as a dependency in Central. While this is not ideal, it will enable us to avoid publishing the NVD CVSS bundle for every Stackrox release if we preprocess and store it in Google Storage during CI. It will also simplify the communication between Central and Scanner if we execute this process within each Scanner instance.```
